lxd/storage/drivers/driver_zfs_utils.go

```diff
@@ -33,6 +33,11 @@ const (
 
 func (d *zfs) dataset(vol Volume, deleted bool) string {
 	name, snapName, _ := api.GetParentAndSnapshotName(vol.name)
+
+	if vol.volType == VolumeTypeImage && shared.IsTrue(vol.ExpandedConfig("zfs.block_mode")) {
```

Comment:
After chatting with Stephane, I don't think we need to encode the filesystem into the volume name, it should be stored as part of the volume's DB record and passed around in volume config. This is what happens in the LVM driver for instance.

See https://github.com/lxc/lxd/blob/master/lxd/storage/backend_lxd.go#L3075-L3086

However because the ZFS driver will support both BlockBacking true and false, we'll need to change that `b.Driver().Info()` implementation in the ZFS driver to something more dynamic (so that it returns the correct value based on config), similar to https://github.com/lxc/lxd/blob/master/lxd/storage/drivers/driver_lvm.go#L88

Reply:
> However because the ZFS driver will support both BlockBacking true and false, we'll need to change that `b.Driver().Info()` implementation in the ZFS driver to something more dynamic (so that it returns the correct value based on config), similar to https://github.com/lxc/lxd/blob/master/lxd/storage/drivers/driver_lvm.go#L88

That won't work because `zfs.block_mode` is per-volume, and `Info()` is pool-based. So `volume.zfs.block_mode` can be `false` for the pool whereas `zfs.block_mode` can be `true` for an instance. In this case `Info().BlockBacking` would falsely return `false`.

Reply:
You're correct that any volume specific operations should use the `zfs.block_mode` setting in the DB config for that volume.

But my link was to `EnsureImage` backendLXD function (https://github.com/lxc/lxd/blob/master/lxd/storage/backend_lxd.go#L3075-L3086) which is for creating image volumes using the current storage pool settings.

That is why I thought we could use the `Info().BlockBacking` option.

When the filesystem or the block backing option changes on the pool, the next time a new instance is created using that image, the backendLXD logic in that function would then detect the filesystem has changed (or is no longer present if block mode has been turned off) and then delete the old image volume and re-create a new image volume.

Now, my understanding of ZFS is that you cannot delete a dataset volume if its used as the origin of other volumes.
This is different to LVM where things are quite a bit easier in that department.

But I think that our ZFS driver already has support for detecting in-use volumes and renaming them with `-deleted` on right?

Reply:
I experimented with unsuffixed images a bit, and found an issue:

```
$ lxc launch testimage c1
  -> creates zfs/images/<fingerprint>

$ lxc storage volume rm zfs images/<fingerprint>
  -> renames zfs/images/<fingerprint> to zfs/deleted/images/<fingerprint>

$ lxc storage set zfs volume.zfs.block_mode=true

$ lxc launch testimage c2
  -> creates zfs/images/<fingerprint>

$ lxc storage volume rm zfs images/<fingerprint>
  -> fails to rename zfs/images/<fingerprint> to zfs/deleted/images/<fingerprint> as it already exists
```

This would succeed if we had suffixes, so I'm in favour of keeping them.

Reply:
> Can we rename with a uuid or something?

I'm sorry, I not following. What exactly do you mean? Can you give an example?

Reply:
We basically need to be able to call DeleteVolume() for an image and have it always succeed, even if its in use.

Reply:
Ceph's `deleteVolume()` function has the concept of zombies, which I think may be useful, as seemingly Ceph has the same issue as ZFS in that you cannot delete a volume that is used as an origin for another volume.

Reply:
Aren't Ceph's zombies basically the same concept as zfs/deleted/...? They are all just renamed images.

Reply:
Yes they are, although Ceph's implementation allows for multiple of the same image (through the use of UUIDs) where as the ZFS doesn't apparently.